dm/syncer/syncer.go

```diff
@@ -2357,26 +2357,35 @@ func (s *Syncer) handleQueryEvent(ev *replication.QueryEvent, ec eventContext, o
 		sourceTbls:      make(map[string]map[string]struct{}),
 		eventStatusVars: ev.StatusVars,
 	}
-	qec.p, err = event.GetParserForStatusVars(ev.StatusVars)
-	if err != nil {
-		log.L().Warn("found error when get sql_mode from binlog status_vars", zap.Error(err))
-	}
 
-	stmt, err := parseOneStmt(qec)
-	if err != nil {
-		// return error if parse fail and filter fail
+	defer func() {
+		if err == nil {
+			return
+		}
+		// return error if filter fail
 		needSkip, err2 := s.skipSQLByPattern(qec.originSQL)
 		if err2 != nil {
-			return err2
+			err = err2
+			return
 		}
 		if !needSkip {
-			return err
+			return
 		}
-		// don't return error if parse fail and filter success
+		// don't return error if filter success
```

Comment:
could you explain why you use defer here ?, old logic is parse -> filter -> track
but now is parse -> handle -> filter 🤔

how about just mute spefic err in https://github.com/pingcap/tiflow/blob/1df27c666ac4266e88a6ca095c4edcb62b12a610/dm/syncer/syncer.go#L2374

Reply:
old logic is

1. parse fail
2. skipSQLByPattern
	1. return err if skip fail
	2. record something if skip success

new logic is

1. for all error (parse fail + DML node + split fail + process fail + ...)
2. skipSQLByPattern
	1. return err if skip fail
	2. record something if skip success

Reply:
yes, but this looks like use defer to impl `try/exception`😂 i konw it can solve #4173, but this looks little strange to me

Reply:
In fact, I don't know why we don't `skipSQLByPattern` at beginning directly, but at later operation failed. @lance6716

Reply:
1. user can configure filter to avoid some ddl being executed downstream
2. user can configure filter to avoid parse error, and we have some built-in filters
3. schema tracker should track every ddl except for the one that will cause error

any implementation is ok

Reply:
> In fact, I don't know why we don't `skipSQLByPattern` at beginning directly, but at later operation failed. @lance6716

> schema tracker should track every ddl except for the one that will cause error

yes, it is order to track every ddl except for the one that will cause error. It's easy to forget. I will add some comment.

Reply:
> yes, but this looks like use defer to impl `try/exception`😂 i konw it can solve #4173, but this looks little strange to me

Indeed. But I haven't thought of a better way temporarily.

Reply:
i thought it again,don't have better suggesion either, the problem is we want to mute err that caused by sql which is user want to skipp,but we also want to track these err 😂

so please add more comment in defer func to explain why we need handle err in this way